### ISSUE TITLE
Support for multiple keys

### ISSUE THREAD

**Issue by oobles** [snazy/ohc#51]
A common implementation to support multiple keys is to create a single object which combines a couple of keys.  ie.

class CombinedKey {
   public PrimaryKey primaryKey;
   public SecondaryKey secondaryKey;
  ...
}

That works fine, but it results in a lot of heap allocation if you're attempting to do high performance (increases GC pressure. Although I do wonder if keeping it short lived enough would keep the object effectively on the stack).

Looking at the linked implementation, the first action most of the methods perform is to serialize the Key object to a KeyBuffer.  Would it be difficult to allow var args for the key and pass these through to the KeySerializer?  Instead of CombinedKey as above you could then use:

map.get(primaryKey, secondaryKey);

I've just been refactoring some other on-heap code to pull apart a combined key like above to reduce the number of objects being allocated.  It would be nice to carry this through to the off heap cache.

**Comment by snazy** [comment]
Hm, I don't think this would work - think of all the methods that take more than one or two args.

Garbage wise, the price is the composite key object (the `CombinedKey` in your example), the internal `KeyBuffer` that takes the serialized representation and a `ByteBuffer` for key serialization.
Gernerally, all these objects should never "escape" make it into the old generation.

I've spent some time a couple of months ago to see whether the current `KeyBuffer` approach could be improved but unfortunately I've found nothing that works better (i.e. paying the price for these objects + GC was still the best). JMH runs have proven that all the approaches are worse.

**Comment by oobles** [comment]
As I just mentioned in issue #49, we've been using a combination of an on-heap cache and off-heap cache.  We've wanted to keep a rather large on-heap cache to reduce serialization overhead.  However this eventually pushes objects into old-gen which can cause longer gc pauses.  So not really an OHC issue.  I've been able to remove all the combined keys for on-heap caches by building my own variants of on-heap caches, so thought I'd see if I can extend this to off-heap.

I've done a very quick initial experiment forking ohc at github.com/future-grid/ohc to see if I could initially separate the map functionality from the off-heap code.  It hasn't been too difficult to separate (although it's currently broken). Once I get the base working again, I'll add a OHTableBuilder and OHTable interface with two keys.  Very experimental at this stage, but still working through the code.  At this stage, the only issue I've got is where keySerialization is called multiple times in a method.  I'm going to try and do a copy keyBuffer instead of calling keySerialization again.

Feel free to close this issue.  If you're interested I'll let you know if/when I've got something working.

**Comment by oobles** [comment]
I spent a bit of time reading through the OHC code to understand the memory layout and design.  I think I've got a good handle on it now.  One thing I was wondering regarding with the KeyBuffer/ByteBuffer allocation.  Did you ever try doing a fixed set of key buffers allocated per thread?  I'm thinking along the lines of a linear probing into a fixed array of key buffers with a lock on each entry when in use.  This could potentially get rid of any allocations for key objects.  Maybe this was one of the things you tried with JMH and discounted?

**Comment by oobles** [comment]
I've done some further experimentation with the OHC code.  I created an OffHeapMap class which implements the normal Map interface to test out some ideas.  This is currently a non-synchronized implementation to test the following ideas:
   - Use a keyScratchPad to serialize the key and then use that buffer internally for key comparisons, etc.
   - Re-use a DataInput/DataOutput stream object held by the cache to write directly to off-heap.  Using DataInput/Output so implementer can benefit from putLong type calls instead of translating everything to bytes.
   - Use a memory manager class to allocate memory on page alignment and use a follow on pointer to allow DataOutput stream to not require upfront knowledge of the length of the output.  Will also add an option for a fixed value size too.
   - Use no allocations and keep interface logic simple enough so I can in future extend to multiple key options.

Currently a work in progress, but thought you might be interested.

**Comment by oobles** [comment]
After some more experimentation we decided to completely rewrite our own off heap cache.  So closing this.  

In case you're interested, we used the following design.  Instead of allocating individual items in memory, we serialized entries directly to a directly allocated ByteBuffer.  This allows keeping it Java friendly (we do use Unsafe to align the byte buffer, although it isn't strictly needed) and lets serialization occur without knowing the size upfront.  We keep an array of ByteBuffers which when full, use a compaction task to clear out any unused or expired entries.  This removes the need for the LRU queue.   These MapStore objects deal only with serialized keys, so we can wrap other facades for multi-key lookups.  The final external API contains multiple MapStore segments to with simple synchronized sections to allow access from multiple threads.
